### ISSUE TITLE
Table/Dataframe viewer improvements

### ISSUE THREAD

**Issue by tlambert03** [napari/magicgui#348]
@kevinyamauchi's wish list from 

**high priority"**
- [ ] Inputs: pandas DataFrame and a dict of numpy arrays
- [ ] Outputs:  pandas DataFrame and a dict of numpy arrays
- [ ] Events: value changed, cell selected, row selected, column selected, others?
- [ ] Style: I don't have super concrete thoughts here, but I think it should be easy for the user and devs to change the styling (e.g., fonts and font sizes) since the size of the tables and table values are going to vary widely. For numeric values, I think it would be useful to specify the precision to avoid super wide float columns (perhaps similar to how it is done in formatted strings).

**"nice to have" (in rough order of my priority)**
- [ ] allow the columns to have types (https://github.com/napari/magicgui/issues/216). bonus points for being able to attach a validator function.
- [ ] ability to "split" (e.g., keep the header row in view while the rest of the table scrolls)
- [ ] allow colormaps to be attached to columns to color cells by value
- [ ] lazy loading
- [ ] a "search bar" the user can use to query the table

**Comment by psobolewskiPhD** [comment]
@tlambert03 related to this, is there any way to get the column headings to be selected and copied along with the data?
I'm also trying to sort out how to make the Table read-only.


> For numeric values, I think it would be useful to specify the precision to avoid super wide float columns (perhaps similar to how it is done in formatted strings). (https://github.com/napari/magicgui/pull/301?)

This can be ticked off I think. At least for floats, I see scientific notation in the Table. But when copied the full precision is there.

A feature to add perhaps, would be a button to save to csv!

**Comment by tlambert03** [comment]
> get the column headings to be selected and copied along with the data?

can you remind me what is meant by "copy" in this case? which method are you using to copy the data?

> Would be nice to have a .editable or something though.

totally agree, would welcome PR

> This can be ticked off I think

thanks!

> a button to save to csv!

agreed!  added to the wishlist

**Comment by psobolewskiPhD** [comment]
> can you remind me what is meant by "copy" in this case? which method are you using to copy the data?

**Comment by tlambert03** [comment]
i see...  it looks like we don't currently have any special handling when one just selects cells and copies them to clipboard in https://github.com/pyapp-kit/magicgui/blob/main/src/magicgui/widgets/_table.py ... so this would then be at the Qt level, and just a convenience that we would need to add.  I'll need to look into the QTable code to see what happens when someone selects and copies cells

**Comment by psobolewskiPhD** [comment]
I think I can do the copy headers bit, working on a PR. Turned out to be easy to implement, just trying to get the test to work!

On the editable/read_only front: it already exists!
https://github.com/pyapp-kit/magicgui/blob/10841894276166ad003a11ae7301a0e181e81e7a/src/magicgui/widgets/bases/_mixins.py#L29

It's just not obvious it exists from the docs:
https://pyapp-kit.github.io/magicgui/api/widgets/Table/

**Comment by psobolewskiPhD** [comment]
@tlambert03 So I don't think QTableWidget has a concept of headers being selected. It's easy enough to use:
```
headers = [
                self.horizontalHeaderItem(c).text()
                for c in range(sel.leftColumn(), sel.rightColumn() + 1)
            ]
lines.append("\t".join(headers))
```
in 
https://github.com/pyapp-kit/magicgui/blob/10841894276166ad003a11ae7301a0e181e81e7a/src/magicgui/backends/_qtpy/widgets.py#L1382

The issue is that if someone was copy/pasting between Table widgets then copying the headers will be bad for them, because they will get an extra row.
Ironically tests pass, because the paste test specifies values and doesn't use copy.

**Comment by psobolewskiPhD** [comment]
Yeah it's 100% a GUI issue, spurred by an end user. I will shelve it for now. Plenty of workaround for the specific use case.
Ultimately, a button to copy the whole table (with headers) would be the way to go.
